crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -0,0 +1,892 @@
+                        BlockStatus::Valid => PayloadStatusEnum::Valid,
+                        BlockStatus::Accepted => PayloadStatusEnum::Accepted,
+                        BlockStatus::Disconnected => PayloadStatusEnum::Syncing,
+                    };
+                    PayloadStatus::new(status, latest_valid_hash)
+                }
+                Err(error) => {
+                    let latest_valid_hash =
+                        matches!(error, Error::Execution(ExecutorError::BlockPreMerge { .. }))
+                            .then_some(H256::zero());
+                    PayloadStatus::new(
+                        PayloadStatusEnum::Invalid { validation_error: error.to_string() },
+                        latest_valid_hash,
+                    )
+                }
+            }
+        } else {
+            PayloadStatus::from_status(PayloadStatusEnum::Syncing)
+        }
+    }
+
+    /// Returns the next pipeline state depending on the current value of the next action.
+    /// Resets the next action to the default value.
+    fn next_pipeline_state(
+        &mut self,
+        pipeline: Pipeline<DB, U>,
+        tip: H256,
+    ) -> PipelineState<DB, U> {
+        let next_action = std::mem::take(&mut self.next_action);
+        if next_action.should_run_pipeline() {
+            trace!(target: "consensus::engine", ?tip, "Starting the pipeline");
+            PipelineState::Running(pipeline.run_as_fut(self.db.clone(), tip))
+        } else {
+            PipelineState::Idle(pipeline)
+        }
+    }
+
+    /// Attempt to restore the tree with the finalized block number.
+    /// If the finalized block is missing from the database, trigger the pipeline run.
+    fn restore_tree_if_possible(
+        &mut self,
+        finalized_hash: BlockHash,
+    ) -> Result<(), reth_interfaces::Error> {
+        match self.db.view(|tx| tx.get::<tables::HeaderNumbers>(finalized_hash))?? {
+            Some(number) => self.blockchain_tree.restore_canonical_hashes(number)?,
+            None => self.pipeline_run_needed(),
+        };
+        Ok(())
+    }
+}
+
+/// On initialization, the consensus engine will poll the message receiver and return
+/// [Poll::Pending] until the first forkchoice update message is received.
+///
+/// As soon as the consensus engine receives the first forkchoice updated message and updates the
+/// local forkchoice state, it will launch the pipeline to sync to the head hash.
+/// While the pipeline is syncing, the consensus engine will keep processing messages from the
+/// receiver and forwarding them to the blockchain tree.
+impl<DB, U, C, EF> Future for BeaconConsensusEngine<DB, U, C, EF>
+where
+    DB: Database + Unpin + 'static,
+    U: SyncStateUpdater + Unpin + 'static,
+    C: Consensus + Unpin,
+    EF: ExecutorFactory + Unpin + 'static,
+{
+    type Output = Result<(), BeaconEngineError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
```

Comment:
this logic seems ok:

drain all messages
poll pipeline
repeat if pipeline not pending